### PR TITLE
Finetune hubert

### PR DESCRIPTION
**Pull request by marcoyang1998**
This recipe implements Hubert transducer model. It supports finetuning a pretrained Hubert model with custom vocabulary using pruned rnnt loss. To use this recipe, you need fairseq as dependency. The finetuning setup (learning rate, optimizer, scheduler, ...) is the same as the original Hubert paper.

### CODE REVIEW

**Comment by marcoyang1998** [comment]
Here are some finetuning results on 960h:
| model name | test-clean  | test-other |
|--------------| ------------- | ------------- |
|Hubert base | 2.82  | 7.09 |
|Hubert large | 1.90  | 3.94 |

Models are trained using BPE500, WERs are obtained using modified beam search.

**Comment by csukuangfj** [comment]
Could you also update `README.md` and `RESULTS.md`?

**Comment by marcoyang1998** [comment]
> Could you also update `README.md` and `RESULTS.md`?

Just did it.

**Comment by desh2608** [comment]
Perhaps some of the files in `finetune_hubert_transducer` can be removed or converted to soft links (such as `asr_datamodule.py`)?

File: egs/librispeech/ASR/RESULTS.md [hunk 1]
```diff
@@ -1077,6 +1077,53 @@ results at:
...
+    --hubert-subsample-output 1 \
+    --hubert-subsample-mode concat_tanh \
+    --encoder-dim 1024 \
+    --enable-spec-aug 0 \
+    --TSA-init-lr 5e-7 \
+    --TSA-warmup-lr 3e-5 \
+    --TSA-end-lr 1.5e-6 \
+    --TSA-total-steps 320000 \
+    --enable-musan 1
+```
```
Comment (csukuangfj) [inline egs/librispeech/ASR/RESULTS.md:1124]: Could you also upload the pretrained model, decoding results, tensorboard logs to huggingface and place a link here?

Comment (marcoyang1998) [reply egs/librispeech/ASR/RESULTS.md:1124]: I will do that soon.

Comment (marcoyang1998) [reply egs/librispeech/ASR/RESULTS.md:1124]: The huggingface link is here https://huggingface.co/marcoyang/icefall_finetuned_hubert

Comment (csukuangfj) [reply egs/librispeech/ASR/RESULTS.md:1124]: Could you place the hugging face link in RESULTS.md, like what we are doing for other recipes?

**Comment by ezerhouni** [comment]
@marcoyang1998 Looks pretty nice !
@marcoyang1998 and @csukuangfj tell me if you need a hand for closing this PR

File: egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+            choices=["concat", "concat_relu", "concat_tanh", "avgpooling"],
+        )
+
+    def normalise_input(
+        self, x: torch.Tensor, x_lens: torch.Tensor
+    ) -> torch.Tensor:
+        """Normalize a batch of audio samples to zero mean unit variance
+
+        Args:
+            x (torch.Tensor): a batch of audio samples
```
Comment (csukuangfj) [inline egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py:153]: Could you add more details about x and x_lens? E.g., the shape of `x` and `x_lens`; is `x` right padded with 0s for some uttterances?

File: egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+            x_lens (torch.Tensor): # audio samples
+
+        Returns:
+            torch.Tensor: normalized audio samples
+        """
+        row_mean = (x.sum(dim=-1) / x_lens).view(x.size(0), 1)
```
Comment (csukuangfj) [inline egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py:159]:
```suggestion
        row_mean = x.sum(dim=-1, keepdim=True) / x_lens
```

File: egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+        row_std = (
+            torch.tensor([sample[sample != 0.0].std() for sample in x])
```
Comment (csukuangfj) [inline egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py:161]: Does `0.0`  have a specialized meaning here? Shall we use the length of `sample` before padding here?

File: egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+            .view(x.size(0), 1)
+            .to(x.device)
+        )
+        for i in range(x.size(0)):
+            x[i, : x_lens[i]] = (x[i, : x_lens[i]] - row_mean[i]) / row_std[i]
```
Comment (csukuangfj) [inline egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py:166]: Shall we remove `row_std` and replace this line with
```suggestion
            x[i, : x_lens[i]] = (x[i, : x_lens[i]] - row_mean[i]) / x[i, : x_lens[i]].std()
```

**Comment by marcoyang1998** [comment]
Sorry, I may be slow to respond during the holiday. Will catch up on all of this after the holiday!

File: egs/librispeech/ASR/RESULTS.md [hunk 6]
```diff
@@ -1077,6 +1077,55 @@ results at:
...
+See <https://github.com/k2-fsa/icefall/pull/588>
+
+A pre-trained Hubert model finetuned with pruned RNN-T loss. Note that fairseq issued
+three versions of [Hubert](https://arxiv.org/abs/2106.07447) of different sizes (base, large, xlarge).
+The first two versions are tested whereas the last one is not due to GPU memory constraint.
+
+The models are finetuned base on the instructions given in [Hubert](https://arxiv.org/abs/2106.07447)
+and [Wav2vec 2.0](https://arxiv.org/abs/2006.11477). Models are trained using Adam optimizer with
+a tri-state lr scheduler. The 960h models are trained for 320k updates using a vocab of 500. The WERs are listed below:
+| model name  | test-clean | test-other |
```
[the remark this answers is not in the document]
Comment (marcoyang1998) [reply egs/librispeech/ASR/RESULTS.md:1093]: Just did it.

File: egs/librispeech/ASR/RESULTS.md [hunk 7]
```diff
@@ -1077,6 +1077,55 @@ results at:
...
+| ----------- |------------|------------|
+| Hubert base |   2.82     | 7.09|
+| Hubert large|   1.93     | 3.93|
+
+you can find the pre-trained model, tensorboard logs and decoding results in:
+<https://huggingface.co/marcoyang/icefall_finetuned_hubert>
```
[the remark this answers is not in the document]
Comment (marcoyang1998) [reply egs/librispeech/ASR/RESULTS.md:1099]: No, it's not generated from export.py. I will replace the model in huggingface repo.

File: egs/librispeech/ASR/RESULTS.md [hunk 8]
```diff
@@ -1077,6 +1077,55 @@ results at:
...
+
+./pruned_transducer_stateless6/train.py \
+    --world-size 6\
+    --num-epochs 25 \
+    --exp-dir finetune_hubert_transducer/exp_960h_HubertL \
+    --full-libri 1 \
+    --max-duration 80 \
+    --bpe-model data/lang_bpe_500/bpe.model \
+    --input-strategy AudioSamples \
+    --hubert-model-dir /ceph-data4/yangxiaoyu/pretrained_models/hubert_large_ll60k.pt \
```
[the remark this answers is not in the document]
Comment (marcoyang1998) [reply egs/librispeech/ASR/RESULTS.md:1112]: Yes, it's downloaded from fairseq. I will add a download function for that.

File: egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py (hunk 4 above)
Comment (marcoyang1998) [reply egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py:161]: No, it doesn't. Your are right, I am changing it.

Comment (marcoyang1998) [reply egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py:161]: No, it doesn't. Your are right, I am changing it.

File: egs/librispeech/ASR/finetune_hubert_transducer/asr_datamodule.py [hunk 9]
```diff
@@ -0,0 +1,455 @@
+# Copyright      2021  Piotr Żelasko
```
[the remark this answers is not in the document]
Comment (marcoyang1998) [reply egs/librispeech/ASR/finetune_hubert_transducer/asr_datamodule.py]: Done

File: egs/librispeech/ASR/finetune_hubert_transducer/beam_search.py [hunk 10]
```diff
@@ -0,0 +1,1541 @@
+# Copyright    2021  Xiaomi Corp.        (authors: Fangjun Kuang)
```
[the remark this answers is not in the document]
Comment (marcoyang1998) [reply egs/librispeech/ASR/finetune_hubert_transducer/beam_search.py:1]: Removed since not used

File: egs/librispeech/ASR/finetune_hubert_transducer/decoder.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
+# Copyright    2021  Xiaomi Corp.        (authors: Fangjun Kuang)
```
[the remark this answers is not in the document]
Comment (marcoyang1998) [reply egs/librispeech/ASR/finetune_hubert_transducer/decoder.py:1]: Done

File: egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py [hunk 12]
```diff
@@ -0,0 +1,260 @@
+#!/usr/bin/env python3
+# Copyright    2021  Xiaomi Corp.        (authors: Xiaoyu Yang)
```
[the remark this answers is not in the document]
Comment (marcoyang1998) [reply egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py:2]: Done

File: egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+    def __init__(
+        self,
+        model_dir: str,
+        output_size: int = 768,
+        freeze_finetune_updates: int = 8000,
+        mask_prob: float = 0.65,
+        mask_channel_prob: float = 0.5,
+        mask_channel_length: int = 64,
+        subsample_output: bool = False,
+        subsample_mode: str = None,
```
[the remark this answers is not in the document]
Comment (marcoyang1998) [reply egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py:37]: Done

File: egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py (hunk 2 above)
Comment (marcoyang1998) [reply egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py:153]: Done

**Comment by marcoyang1998** [comment]
I updated the huggingface repo. Everything should be ready now.

**Comment by csukuangfj** [comment]
> I updated the huggingface repo. Everything should be ready now.

Thanks! Will look at it later today.

File: egs/librispeech/ASR/RESULTS.md [hunk 14]
```diff
@@ -1077,6 +1077,76 @@ results at:
...
+### LibriSpeech BPE training results (Hubert Transducer)
+
+[finetune_hubert_transducer](./finetune_hubert_transducer)
+
+See <https://github.com/k2-fsa/icefall/pull/588>
+
+A pre-trained Hubert model finetuned with pruned RNN-T loss. Note that fairseq issued
+three versions of [Hubert](https://arxiv.org/abs/2106.07447) of different sizes (base, large, xlarge).
+The first two versions are tested whereas the last one is not due to GPU memory constraint. Before
+trying this recipe, you may need to first download the fairseq pretrained HuBert models. See (./finetune_hubert_transducer/download.sh) for instructions.
```
Comment (csukuangfj) [inline egs/librispeech/ASR/RESULTS.md:1089]:
```suggestion
trying this recipe, you may need to first download the fairseq pretrained HuBert models. See [./finetune_hubert_transducer/download.sh](./finetune_hubert_transducer/download.sh) for instructions.
```

File: egs/librispeech/ASR/finetune_hubert_transducer/asr_datamodule.py [hunk 15]
```diff
@@ -0,0 +1 @@
+/ceph-data4/yangxiaoyu/softwares/icefall_development/icefall_forked/egs/librispeech/ASR/tdnn_lstm_ctc/asr_datamodule.py
```
Comment (csukuangfj) [inline egs/librispeech/ASR/finetune_hubert_transducer/asr_datamodule.py:1]: Could you use a relative path?
You can refer to what we are doing for other folders with symlinks.

File: egs/librispeech/ASR/finetune_hubert_transducer/conformer.py [hunk 16]
```diff
@@ -0,0 +1,1848 @@
+#!/usr/bin/env python3
```
Comment (csukuangfj) [inline egs/librispeech/ASR/finetune_hubert_transducer/conformer.py:1]: Could you replace it with a symlink to pruned_transducer_stateless5/conformer.py?

Also for the other files that you copied directly, e.g., `joiner.py`.

(I see that you have already done this for asr_datamodule.py)

File: egs/librispeech/ASR/finetune_hubert_transducer/decoder.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (csukuangfj) [reply egs/librispeech/ASR/finetune_hubert_transducer/decoder.py:1]: Have you pushed the changes? It is still not changed.

**Comment by marcoyang1998** [comment]
I need to update pretrained.py as hubert requires waveform as input.

File: egs/librispeech/ASR/finetune_hubert_transducer/decode.py [hunk 17]
```diff
@@ -0,0 +1,608 @@
...
+        "using --epoch. If True, it would decode with the averaged model "
+        "over the epoch range from `epoch-avg` (excluded) to `epoch`."
+        "Actually only the models with epoch number of `epoch-avg` and "
+        "`epoch` are loaded for averaging. ",
+    )
+
+    parser.add_argument(
+        "--exp-dir",
+        type=str,
+        default="pruned_transducer_stateless6/exp",
```
Comment (csukuangfj) [inline egs/librispeech/ASR/finetune_hubert_transducer/decode.py:131]:
```suggestion
        default="finetune_hubert_transducer/exp_960h",
```

File: egs/librispeech/ASR/finetune_hubert_transducer/decode.py [hunk 18]
```diff
@@ -0,0 +1,608 @@
...
+        fast_beam_search""",
+    )
+
+    parser.add_argument(
+        "--max-states",
+        type=int,
+        default=8,
+        help="""Used only when --decoding-method is
+        fast_beam_search""",
+    )
```
Comment (csukuangfj) [inline egs/librispeech/ASR/finetune_hubert_transducer/decode.py:185]:
```suggestion
```

These arguments are for fast_beam_search. If this script does not support it, better to remove them.

File: egs/librispeech/ASR/finetune_hubert_transducer/decode.py [hunk 19]
```diff
@@ -0,0 +1,608 @@
...
+
+    return parser
+
+
+def decode_one_batch(
+    params: AttributeDict,
+    model: nn.Module,
+    sp: spm.SentencePieceProcessor,
+    batch: dict,
+    decoding_graph: Optional[k2.Fsa] = None,
```
Comment (csukuangfj) [inline egs/librispeech/ASR/finetune_hubert_transducer/decode.py:213]:
```suggestion
```

File: egs/librispeech/ASR/finetune_hubert_transducer/decode.py [hunk 20]
```diff
@@ -0,0 +1,608 @@
...
+        The neural model.
+      sp:
+        The BPE model.
+      batch:
+        It is the return value from iterating
+        `lhotse.dataset.K2SpeechRecognitionDataset`. See its documentation
+        for the format of the `batch`.
+      decoding_graph:
+        The decoding graph. Can be either a `k2.trivial_graph` or HLG, Used
+        only when --decoding_method is fast_beam_search.
```
Comment (csukuangfj) [inline egs/librispeech/ASR/finetune_hubert_transducer/decode.py:238]:
```suggestion
```

File: egs/librispeech/ASR/finetune_hubert_transducer/decode.py [hunk 21]
```diff
@@ -0,0 +1,608 @@
...
+                    filename_start=filename_start,
+                    filename_end=filename_end,
+                    device=device,
+                )
+            )
+
+    model.to(device)
+    model.eval()
+
+    if params.decoding_method == "fast_beam_search":
```
Comment (csukuangfj) [inline egs/librispeech/ASR/finetune_hubert_transducer/decode.py:568]: By the way, what are the difficulties in supporting fast_beam_search?

File: egs/librispeech/ASR/finetune_hubert_transducer/export.py [hunk 22]
```diff
@@ -0,0 +1,308 @@
...
+
+    cd /path/to/exp_dir
+    ln -s pretrained.pt epoch-9999.pt
+
+    cd /path/to/egs/librispeech/ASR
+    ./finetune_hubert_transducer/decode.py \
+        --exp-dir ./finetune_hubert_transducer/exp_dir \
+        --epoch 9999 \
+        --avg 1 \
+        --max-duration 600 \
```
Comment (csukuangfj) [inline egs/librispeech/ASR/finetune_hubert_transducer/export.py:49]: I thin you have to use
```
--use-averaged-model False
```
